### PR TITLE
[SYCL-MLIR] Add flag to allow relax sycl-opt tweaks

### PR DESCRIPTION
**Pull request by Naghasan**
`sycl-opt` llvm flag prevent a few loop optimizations.

The patch add `-fsycl-relaxed-opt` to clang and `sycl-relaxed-opt` in llvm which allows loop rotate, idom and simplify to run.

### CODE REVIEW

File: llvm/lib/Passes/PassBuilderPipelines.cpp [hunk 1]
```diff
@@ -183,6 +183,9 @@ static cl::opt<bool> EnableGlobalAnalyses(
 static cl::opt<bool>
     SYCLOptimizationMode("sycl-opt", cl::init(false), cl::Hidden,
                          cl::desc("Enable SYCL optimization mode."));
+static cl::opt<bool>
+    SYCLRelaxedOptimizationMode("sycl-relaxed-opt", cl::init(false), cl::Hidden,
+                         cl::desc("Enable SYCL optimization mode."));
```
Comment (whitneywhtsang) [inline llvm/lib/Passes/PassBuilderPipelines.cpp:188]: description need update.

**Comment by sommerlukas** [comment]
> I'm confused. Looks to me that if `-fsycl-relaxed-opt` is used then a number of LLVM opt passes aren't run, an example LICM will be run only if `-fsycl-relaxed-opt` is _not_ used:
> 
> ```
>    if (!SYCLRelaxedOptimizationMode) {
>       LPM1.addPass(LICMPass(PTO.LicmMssaOptCap, PTO.LicmMssaNoAccForPromotionCap,
>                             /*AllowSpeculation=*/false));
>     }
> ```

`sycl-opt` works in a similar fashion, it *disables* some optimizations which might have negative impact on the device driver. 

Because the eventual device target is not known when compiling to SPIR-V (as it is a retargetable IR), some optimizations are often left to the device driver.

**Review by whitneywhtsang — COMMENTED** [review]
Long term we should find out why loop rotate, idom and simplify are run for DPC++ but not SYCL-MLIR. Given that this is short term, I suggest to have a dup of `buildO1FunctionSimplificationPipeline` function for `SYCLRelaxedOptimizationMode` , to reduce merge conflicts and easier to see what is in the pipeline.

**Comment by Naghasan** [comment]
> Long term we should find out why loop rotate, idom and simplify are run for DPC++ but not SYCL-MLIR

`sycl-opt` disable loop rotate, idom and simplify, so they don't run for DPC++, the `sycl-relaxed-opt` enables them (the patch equally affects DPC++ and SYCL-MLIR).

The thing I was looking at it the difference between the 2 generated LLVM IR at the end of the pipeline. Some initialization done by Clang's codegen is done via a straight forward loop doing a memcpy: 1 BB branching onto itself (I can't recall why they don't emit an `llvm.memcpy` instead). What we think happens with our MLIR is polygeist generates an `scf.for`/`affine.for` for the same initializer which is then lowered into a traditional for loop (condition check then branch to loop body) with, I guess, the assumption the loop pipeline cleaning that out.
Because `sycl-opt` blindly disable the loop pipeline, we end up with these structure.

File: llvm/lib/Passes/PassBuilderPipelines.cpp (hunk 1 above)
Comment (Naghasan) [reply llvm/lib/Passes/PassBuilderPipelines.cpp]: Fixed

**Comment by Naghasan** [comment]
````
I'm confused. Looks to me that if -fsycl-relaxed-opt is used then a number of LLVM opt passes aren't run, an example LICM will be run only if -fsycl-relaxed-opt is not used:
````

As Lukas said, works in the same way as `sycl-opt`. The flags doesn't add anything to the pipeline but restrict it (kills the loop pipeline), here I'm selectively disabling some pass of the pipeline.

But I'm being very conservative as I don't want to ending up generating freeze instruction (can't really be handled by the SPIR-V translator)

**Comment by Naghasan** [comment]
@whitneywhtsang I cloned the functions

**Comment by Naghasan** [comment]
Closing as this didn't significantly improved the situation.

**Pull request closed without merging**